libs/@blockprotocol/graph/src/types/subgraph/element-mappings.ts

```diff
@@ -0,0 +1,120 @@
+import { BaseUri, VersionedUri } from "@blockprotocol/type-system";
+
+import { Entity, EntityId, EntityRecordId } from "../entity";
+import {
+  DataTypeWithMetadata,
+  EntityTypeWithMetadata,
+  OntologyTypeRecordId,
+  PropertyTypeWithMetadata,
+} from "../ontology";
+import { EntityValidInterval } from "./edges";
+import { EntityIdAndTimestamp } from "./edges/outward-edge";
+import {
+  DataTypeVertex,
+  EntityTypeVertex,
+  EntityVertex,
+  EntityVertexId,
+  OntologyTypeVertexId,
+  PropertyTypeVertex,
+} from "./vertices";
+
+/**
+ * A utility type that maps various ways of identifying a single (or series of) element(s) of the graph to their
+ * associated types.
+ *
+ * Helpful when creating generic functions that operate over a {@link Subgraph}
+ */
+export type GraphElementIdentifiers<Temporal extends boolean> =
+  | {
+      identifier: VersionedUri | OntologyTypeVertexId | OntologyTypeRecordId;
+      element:
+        | DataTypeWithMetadata
+        | PropertyTypeWithMetadata
+        | EntityTypeWithMetadata;
+      vertex: DataTypeVertex | PropertyTypeVertex | EntityTypeVertex;
+    }
+  | {
+      identifier: BaseUri;
+      element:
+        | DataTypeWithMetadata[]
+        | PropertyTypeWithMetadata[]
+        | EntityTypeWithMetadata[];
+      vertex: DataTypeVertex[] | PropertyTypeVertex[] | EntityTypeVertex[];
+    }
+  | {
+      identifier: EntityIdAndTimestamp | EntityVertexId | EntityRecordId;
+      element: Entity<Temporal>;
+      vertex: EntityVertex<Temporal>;
+    }
+  | {
+      identifier: EntityId | EntityValidInterval;
+      element: Entity<Temporal>[];
+      vertex: EntityVertex<Temporal>[];
+    };
```

Comment:
That's a very interesting question. I'm not sure if it's worth separating them out in this type (it could cause problems with other advanced types like `Extract`).

I don't think it'd necessarily be an error for a user to try and use these things on a non-temporal subgraph. _At the moment_ revisionId's in a non-temporal subgraph are still timestamps, they're just mostly meaningless, and there won't be multiple revisions.

Prior to the past two PRs for example, the stdlib relied on passing in the current timestamp to get an entity (only one revision as we're talking non-temporal) by its `EntityId`.

I can add a todo in the code if you'd like though, this might be worth us investigating later. One of the things I think this type _could_ be helpful for, is forcing us to make it ergonomic to accept any sort of identifier in the stdlib. Right now, some of the `get*` methods (such as `getPropertyTypeByVertexId`) take in a specific kind of identifier, which requires the user to either select the right function (if it exists), or transform an identifier they just got from a different part of the stdlib, into another identifier. We _could_ make all of the `getPropertyType*` methods take a generic "identifier" thing from this mapping object, and then you'd be forced to handle all of them, which seems like good UX. cc @CiaranMn for my intentions long term here.